### DISCUSSION TITLE
What is the rationale for preventing null propagation with method groups?

### DISCUSSION THREAD

**Discussion by jnm2** [dotnet/csharplang#5336] [Q&A]
[SharpLab](https://sharplab.io/#v2:C4LgbgpgTglgZgTwD4AEBMBGAsAKF+gAgGFcBvXAyglDNAgWQAoBJAERgGcAHAew4EMARgBsIBACadeAkRACUFKuRxVVBAPTqCgGXJiAZQAMBtAGYQBAPJdo/YDygEA5AH5HBAMb8Adl57ACgmL8XFzCMBDiBHYEPNZQ3pE8cFEI1k4AthDAABY8kQDmUDwArlyOimoaWpU1BIDJhBVqKADsElJ8QqLOAHTs3HwQANyNAL64YzhAA===)

```cs
using System;

class C
{
    Action M(IDisposable disposable)
    {
        // ❌ CS0023: Operator '?' cannot be applied to operand of type 'method group'
        //               ↓
        return disposable?.Dispose;
    }
}
```

This is the code that must currently be used instead:
```cs
using System;

class C
{
    Action M(IDisposable disposable)
    {
        return disposable is not null ? disposable.Dispose : null;
    }
}
```

**Comment by alrz** [comment]
>What is the rationale for preventing null propagation with method groups?

Not implemented.

**Reply by jnm2** [reply]
Meaning the compiler doesn't yet comply with the spec, or the spec work wasn't done?

**Reply by alrz** [reply]
The latter of course. I'd be surprised if something as big as this was just *slipped* after seven years.

**Comment by Rekkonnect** [comment]
I had no idea this doesn't work. With a bit of tweaking, I also discovered that `??` on method groups doesn't work either (in such context).

**Reply by jnm2** [reply]
Oh, I remember this. This one didn't super surprise me though, since the `new Action(...)` part would have to be communicated through the `??` operator by target-typing. I guess maybe the same thing is true of the `?.` operator in some sense that I'm not seeing.

**Reply by Rekkonnect** [reply]
Whatever holds true, target type is visible and explicit. Method group should be bound as early as possible, enabling use of `?.` and `??`. Almost everyone would expect this behavior to be valid.

**Reply by jnm2** [reply]
I don't think so actually in the case of `??`. Target typing doesn't normally get communicated through the `??` operator to either of its operands:

```cs
class C
{
    C M()
    {
        // ❌ CS8754: There is no target type for 'new()'
        return new() ?? null;
    }
}
```

```cs
class C
{
    C M()
    {
        // ❌ CS0019: Operator '??' cannot be applied to operands of type '<null>' and 'new()'
        return null ?? new();
    }
}
```

So I wouldn't expect target typing to `SomeDelegate` to happen for a method group operand, either.

**Reply by Rekkonnect** [reply]
You're right, target typing for `??` is another proposal on its own, and improving `?.` for method groups should also enable `??` for method groups once `??` target typing is on, or at least closer to that step.

**Comment by 333fred** [comment]
I'd be willing to champion changing this.

**Comment by jnm2** [comment]
https://github.com/dotnet/csharpstandard/issues/654 explores whether the spec actually currently allows this, via target-typed conditional expressions, and `?.` being defined as lowering to a conditional expression. I'm not sure there's been a conclusion yet.
